src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs

```diff
@@ -418,30 +408,82 @@ public async Task<(SolutionUpdateStatus Summary, ImmutableArray<Deltas> Deltas)>
         }
 
         public void DiscardSolutionUpdate()
         {
-            Contract.ThrowIfNull(_editSession);
-
-            var pendingUpdate = Interlocked.Exchange(ref _pendingUpdate, null);
-            Contract.ThrowIfNull(pendingUpdate);
+            var editSession = _editSession;
+            Contract.ThrowIfNull(editSession);
 
+            var pendingUpdate = editSession.RetrievePendingUpdate();
             foreach (var moduleReader in pendingUpdate.ModuleReaders)
             {
                 moduleReader.Dispose();
             }
         }
 
-        public async Task<LinePositionSpan?> GetCurrentActiveStatementPositionAsync(ActiveInstructionId instructionId, CancellationToken cancellationToken)
+        public async Task<ImmutableArray<ImmutableArray<(LinePositionSpan, ActiveStatementFlags)>>> GetBaseActiveStatementSpansAsync(ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
+        {
+            var editSession = _editSession;
+            if (editSession == null)
+            {
+                return default;
+            }
+
+            var lastCommittedSolution = editSession.DebuggingSession.LastCommittedSolution;
+            var baseActiveStatements = await editSession.BaseActiveStatements.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            using var _ = ArrayBuilder<ImmutableArray<(LinePositionSpan, ActiveStatementFlags)>>.GetInstance(out var spans);
+
+            foreach (var documentId in documentIds)
+            {
+                if (baseActiveStatements.DocumentMap.TryGetValue(documentId, out var documentActiveStatements))
+                {
+                    var (baseDocument, _) = await lastCommittedSolution.GetDocumentAndStateAsync(documentId, cancellationToken).ConfigureAwait(false);
+                    if (baseDocument != null)
+                    {
+                        spans.Add(documentActiveStatements.SelectAsArray(s => (s.Span, s.Flags)));
+                        continue;
+                    }
+                }
+
+                // Document contains no active statements, or
+                // document has been added, is out-of-sync or a design-time-only document.
+                spans.Add(ImmutableArray<(LinePositionSpan, ActiveStatementFlags)>.Empty);
+            }
+
+            return spans.ToImmutable();
+        }
+
+        public async Task<ImmutableArray<(LinePositionSpan, ActiveStatementFlags)>> GetDocumentActiveStatementSpansAsync(Document document, CancellationToken cancellationToken)
+        {
+            var editSession = _editSession;
+            if (editSession == null)
+            {
+                return default;
+            }
+
+            var lastCommittedSolution = editSession.DebuggingSession.LastCommittedSolution;
+            var (baseDocument, _) = await lastCommittedSolution.GetDocumentAndStateAsync(document.Id, cancellationToken).ConfigureAwait(false);
+            if (baseDocument == null)
```

Comment:
nit for the future: it's sometimes good to doc these in terms of if this is just being cautious or if it's a real world scenario.  i'm guessing it's real world (i.e a new file was added), but it would be helpful to know.

Reply:
The method explains in its doc comment:
```
        /// <summary>
        /// Returns a document snapshot for given <see cref="DocumentId"/> whose content exactly matches
        /// the source file used to compile the binary currently loaded in the debuggee. Returns null
        /// if it fails to find a document snapshot whose content hash maches the one recorded in the PDB.
        /// 
        /// The result is cached and the next lookup uses the cached value, including failures unless <paramref name="reloadOutOfSyncDocument"/> is true.
        /// </summary>        
```